src/com/google/javascript/jscomp/parsing/parser/util/UnicodeMatch.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2020 The Closure Compiler Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.javascript.jscomp.parsing.parser.util;
+
+import java.util.regex.Pattern;
+
+/**
+ * Exposes non-ASCII unicode detection for Javascript identifiers in compliance
+ * with ECMAScript spec. Retains old optimizations for commonly-used characters,
+ * namely ASCII and Latin/Greek charsets. Uses regex for remaining characters
+ * (which would throw errors otherwise).
+ * 
+ * Unicode category patterns with /u flag compiled to vanilla regex with
+ * regexpu: https://github.com/mathiasbynens/regexpu
+ * 
+ * IdentifierStart and IdentifierPart regex patterns based on ECMA's
+ * documentation:
+ * https://github.com/tc39/proposal-regexp-unicode-property-escapes#other-examples
+ * 
+ * const regexIdentifierStart = /[$_\p{ID_Start}]/u;
+ * const regexIdentifierPart = /[$_\u200C\u200D\p{ID_Continue}]/u;
+ */
+public class UnicodeMatch {
+  // Intentional to minimize branches in this code
+  @SuppressWarnings("ShortCircuitBoolean")
+  private static boolean fastPathLetters(char ch) {
+    return (
+      // ASCII
+      ('A' <= ch & ch <= 'Z') |
+      ('a' <= ch & ch <= 'z') |
+      (ch == '_' | ch == '$') |
+      // Latin letters
+      (0x00C0 <= ch & ch <= 0x00D6) |
+      // Latin letters
+      (0x00D8 <= ch & ch <= 0x00F6) |
+      // Latin letters
+      (0x00F8 <= ch & ch <= 0x00FF) |
+      // Latin Barred O
+      (ch == 0x0275) |
+      // Greek uppercase letters
+      (0x0391 <= ch & ch <= 0x03A1) |
+      // Remaining Greek uppercase letters
+      (0x03A3 <= ch & ch <= 0x03A9) |
+      // Greek lowercase letters
+      (0x03B1 <= ch & ch <= 0x03C9)
+    );
+  }
+
+  // Intentional to minimize branches in this code
+  @SuppressWarnings("ShortCircuitBoolean")
+  private static boolean fastPathDigits(char ch) {
+    return '0' <= ch & ch <= '9';
+  }
+
+  public static boolean isJavascriptIdentifierStart(char ch) {
+    return fastPathLetters(ch)
+        || JavascriptIdentifierStart
+              .matcher(Character.toString(ch))
+              .matches();
+  }
+
+  public static boolean isJavascriptIdentifierPart(char ch) {
```

Comment:
I would add a comment here noting that this won't correctly identify supplementary characters (codepoint > 0xFFFF), because they don't fit in a `char`. Fixing this would require additional changes to the parser, which currently uses single-char lookahead.

Reply:
@tjgq You're right - I will update to `int`.